Sources/MongoSwift/BSON/BSONDecoder.swift

```diff
@@ -3,17 +3,74 @@ import Foundation
 /// `BSONDecoder` facilitates the decoding of BSON into semantic `Decodable` types.
 public class BSONDecoder {
 
+    @available(OSX 10.12, *)
+    internal static var iso8601Formatter: ISO8601DateFormatter = {
+        let formatter = ISO8601DateFormatter()
+        formatter.formatOptions = .withInternetDateTime
+        return formatter
+    }()
+
+    /// Enum representing the different options for decoding `Date`s from BSON
+    public enum DateDecodingStrategy {
+        /// Decoding `Date`s stored as 64 bit integers counting the number of milliseconds since January 1, 1970
+        case millisecondsSince1970Int64
+
+        /// Decoding `Date`s stored as 64 bit integers counting the number of seconds since January 1, 1970
+        case secondsSince1970Int64
+
+        /// Decoding `Date`s stored as `Double`s counting the number of seconds since January 1, 1970
+        case millisecondsSince1970
+
+        /// Decoding `Date`s stored as `Double`s counting the number of milliseconds since January 1, 1970
+        case secondsSince1970
+
+        /// Decoding `Date`s by deferring to their default decoding implementation
+        case deferredToDate
+
+        /// Decoding `Date`s represented by ISO8601 formatted strings
+        @available(OSX 10.12, *)
```

Comment:
The `*` means "potential future platforms".

For this specific case, `JSONEncoder` uses `available(macOS 10.12, iOS 10.0, watchOS 3.0, tvOS 10.0, *)`, which covers everything.

When I looked up the source of `ISO8601DateFormatter` though, its own attribute was `@available(OSX 10.12, *)`, which is what I went with. Now that I'm looking again, I can't find that anywhere... 

At any rate, I'll stick with what `JSONDecoder` does, as it matches what the [docs](https://developer.apple.com/documentation/foundation/iso8601dateformatter) say as well.

As far as Linux support goes, I'm not really sure. The availability of this formatter seems dependent upon platform instead of Swift version, which is pretty odd. Linux users could always use `.formatted` with their own formatter to achieve the same results.

Reply:
I tested it on Sam's machine, and the availability condition passes. I guess Linux is either included under `*` or the Linux version of Foundation is macOS versioning aware. Either way, it seems to work fine.